File: backend/app/models/notification.py
from sqlalchemy import Column, Integer, String, Boolean, ForeignKey, DateTime, Text, func
from sqlalchemy.orm import relationship
from ..core.database import Base


class Notification(Base):
    __tablename__ = "notifications"

    notification_id = Column(Integer, primary_key=True, index=True)
    recipient_staff_id = Column(Integer, ForeignKey("staff.staff_id", ondelete="CASCADE"), nullable=False)
    sender_staff_id = Column(Integer, ForeignKey("staff.staff_id", ondelete="SET NULL"))
    notification_type = Column(String(50))  # task_assigned, task_status_changed, shift_assigned
    title = Column(String(255), nullable=False)
    message = Column(Text)
    link_url = Column(String)  # /tasks/123, /shifts
    is_read = Column(Boolean, default=False)
    read_at = Column(DateTime)
    created_at = Column(DateTime, default=func.now())

    # Relationships
    recipient = relationship("Staff", foreign_keys=[recipient_staff_id], back_populates="received_notifications")
    sender = relationship("Staff", foreign_keys=[sender_staff_id], back_populates="sent_notifications")

File: backend/app/Models/Notification.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;

class Notification extends Model
{
    protected $table = 'notifications';
    protected $primaryKey = 'notification_id';
    public $timestamps = true;

    const CREATED_AT = 'created_at';
    const UPDATED_AT = 'updated_at';

    protected $fillable = [
        'recipient_staff_id',
        'sender_staff_id',
        'title',
        'message',
        'type',
        'priority',
        'is_read',
        'read_at',
        'action_url',
        'data',
    ];

    protected $casts = [
        'is_read' => 'boolean',
        'read_at' => 'datetime',
        'data' => 'array',
        'created_at' => 'datetime',
        'updated_at' => 'datetime',
    ];

    public function recipient()
    {
        return $this->belongsTo(Staff::class, 'recipient_staff_id', 'staff_id');
    }

    public function sender()
    {
        return $this->belongsTo(Staff::class, 'sender_staff_id', 'staff_id');
    }

    // Mark as read
    public function markAsRead()
    {
        $this->update([
            'is_read' => true,
            'read_at' => now(),
        ]);
    }

    // Scope for unread notifications
    public function scopeUnread($query)
    {
        return $query->where('is_read', false);
    }

    // Scope for recipient
    public function scopeForRecipient($query, $staffId)
    {
        return $query->where('recipient_staff_id', $staffId);
    }
}
